### ISSUE TITLE
Readiness, Liveness probes always failing

### ISSUE THREAD

**Issue by FaresHerhar** [coder/coder#5577]
> Description

I am installing coder v2 on my Kubernetes cluster, the installation goes well, but the Coder pod is not running, and when I get into the pod's describe, it says that both Readiness and Liveness probes fails, due connection problems.

I am using the following official documentation <https://coder.com/docs/coder-oss/latest/install/kubernetes>.

I tried it on both Minikube, and a baremetal Kubernetes Cluster (kubeadm based installation), and I am having the same problem on both !

> Status

NAME                     READY   STATUS             RESTARTS        AGE
coder-7c9686b56d-g5kvw   0/1     CrashLoopBackOff   6 (2m17s ago)   12m
coder-db-postgresql-0    1/1     Running            0               15m

> k describe pod/coder-7c9686b56d-g5kvw -n coder

  Normal   Scheduled  11m                  default-scheduler  Successfully assigned coder/coder-7c9686b56d-g5kvw to minikube
  Normal   Pulling    11m                  kubelet            Pulling image "ghcr.io/coder/coder:v0.13.6"
  Normal   Pulled     11m                  kubelet            Successfully pulled image "ghcr.io/coder/coder:v0.13.6" in 13.559505282s
  Warning  Unhealthy  11m                  kubelet            Readiness probe failed: Get "http://10.244.0.5:8080/api/v2/buildinfo": dial tcp 10.244.0.5:8080: connect: connection refused
  Warning  Unhealthy  11m (x3 over 11m)    kubelet            Liveness probe failed: Get "http://10.244.0.5:8080/api/v2/buildinfo": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Normal   Killing    11m                  kubelet            Container coder failed liveness probe, will be restarted
  Normal   Created    10m (x2 over 11m)    kubelet            Created container coder
  Normal   Started    10m (x2 over 11m)    kubelet            Started container coder
  Normal   Pulled     10m                  kubelet            Container image "ghcr.io/coder/coder:v0.13.6" already present on machine
  Warning  Unhealthy  108s (x79 over 11m)  kubelet            Readiness probe failed: Get "http://10.244.0.5:8080/api/v2/buildinfo": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

**Comment by bpmct** [comment]
Can you check the logs of the pod? `kubectl logs coder-7c9686b56d-g5kvw`

Also if you are running Coder in an offline environment, you'll need to build a [custom Coder image](https://coder.com/docs/coder-oss/latest/install/offline) that includes Terraform and any providers you plan to use.

**Comment by FaresHerhar** [comment]
For the logs, I can't get anything, since the pod is continually restarting within 10 seconds.
I am not in an offline mode, since it's minikube, but I'll try to build the offline image, and get back to you.

Thanks for the suggestion.

**Comment by FaresHerhar** [comment]
> Can you check the logs of the pod? `kubectl logs coder-<id>`
> 
> Also if you are running Coder in an offline environment, you'll need to build a [custom Coder image](https://coder.com/docs/coder-oss/latest/install/offline) that includes Terraform and any providers you plan to use.

Hello,
I tried the offline installation version, and I am still running on the same issue.
Weird  I know !

I've tested it on both clusters, and still fires the same error !
Has anyone tried to do the installation with the current version ?

> This is the code of the probes

```yaml
          readinessProbe:
            httpGet:
              path: /api/v2/buildinfo
              port: "http"
              scheme: "HTTP"
          livenessProbe:
            httpGet:
              path: /api/v2/buildinfo
              port: "http"
              scheme: "HTTP"
```

I  logged into the pod (I had to extend the probes periods ), and I ran this `curl 0.0.0.0:8080/api/v2/buildinfo`, and it just times out, and I get logged of the pod, because it fires an error.

As I a final note, I checked the environment variables (the postgresql  address is well communicated), and the postgresql  pod is running, I am able to access it.

> K9S Screenshot

![image](https://user-images.githubusercontent.com/102948288/211008768-c66004e5-66ca-4c93-aaa8-f5442d2c7fa0.png)

**Comment by FaresHerhar** [comment]
Problem fixed.
Thanks.

We had an issue with docker permissions, which caused the Liveness & Rediness probes issues.

**Comment by abhinavsingh** [comment]
1) Can someone share what exactly is the issue with docker permissions here?
2) I followed the kubernetes guide to install via helm charts
3) But the pod keeps crashing due to probe failures
4) I have tried both stable and mainline releases
5) Postgres db was not installed via bitmani, but through direct statefulset with replica 1
6) Postgres db is reachable and coder-db-url is setup correctly
7) PS: I changed the default user and db (not coder) for my Postgres, but I doubt this should be an issue here
8) I used netstat and do see port 8080 open on the pod
9) However, curl localhost:8080/healthz or via pod IP simply stucks
10) Nothing was found under termination logs either

Can we get some help here please.  Thank you

```
Events:
  Type     Reason     Age                   From               Message
  ----     ------     ----                  ----               -------
  Normal   Scheduled  4m13s                 default-scheduler  Successfully assigned coder/coder-9bbd694b4-9b5x5 to jaxl-workstation-2
  Normal   Pulling    4m12s                 kubelet            Pulling image "ghcr.io/coder/coder:v2.14.2"
  Normal   Pulled     3m8s                  kubelet            Successfully pulled image "ghcr.io/coder/coder:v2.14.2" in 1m3.631s (1m3.631s including waiting). Image size: 131868133 bytes.
  Warning  Unhealthy  2m42s (x3 over 3m1s)  kubelet            Liveness probe failed: Get "http://10.244.192.2:8080/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Normal   Killing    2m42s                 kubelet            Container coder failed liveness probe, will be restarted
  Normal   Created    101s (x2 over 3m8s)   kubelet            Created container coder
  Normal   Started    101s (x2 over 3m8s)   kubelet            Started container coder
  Normal   Pulled     101s                  kubelet            Container image "ghcr.io/coder/coder:v2.14.2" already present on machine
  Warning  Unhealthy  99s (x14 over 3m6s)   kubelet            Readiness probe failed: Get "http://10.244.192.2:8080/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

**Comment by abhinavsingh** [comment]
jFYI, I tried a few things and here are my observations:

1) I disabled probes and started to see following logs

```console
Started HTTP listener at http://0.0.0.0:8080

View the Web UI: https://coder.<domain.com>
2024-08-23 04:16:21.718 [erro]  connect to postgres failed ...
    error= migrate up:
               github.com/coder/coder/v2/cli.ConnectToPostgres
                   /home/runner/work/coder/coder/cli/server.go:2157
             - up:
               github.com/coder/coder/v2/coderd/database/migrations.UpWithFS
                   /home/runner/work/coder/coder/coderd/database/migrations/migrate.go:82
             - timeout: can't acquire database lock
Encountered an error running "coder server", see "coder server --help" for more information
error: connect to postgres: connect to postgres: migrate up: up: timeout: can't acquire database lock
```

2) It crashed and restarted, then I saw these logs

```console
Started HTTP listener at http://0.0.0.0:8080

View the Web UI: https://coder.<domain.com>
2024-08-23 04:23:59.471 [info]  pubsub: pubsub dialing postgres  network=tcp  address=postgres-coder-0.postgres-coder.coder.svc.cluster.local:5432  timeout_ms=0
2024-08-23 04:23:59.502 [info]  pubsub: pubsub postgres TCP connection established  network=tcp  address=postgres-coder-0.postgres-coder.coder.svc.cluster.local:5432  timeout_ms=0  elapsed_ms=30
2024-08-23 04:23:59.514 [info]  pubsub: pubsub connected to postgres
2024-08-23 04:23:59.514 [info]  pubsub: pubsub has started

==> Logs will stream in below (press ctrl+c to gracefully exit):
2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine network map  network_map="netmap: self: [6O1/Y] auth=machine-unknown u=? [fd7a:115c:a1e0:4312:8465:85bc:3ed5:4a10/128]\n"
2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine filter  filter="&{logf:0x136d7c0 local:0xc000b7ebb8 logIPs:0xc000b7ebd0 matches4:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:281470681743360} z:0xc00012c078} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:281470681743360} z:0xc00012c078} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] matches6:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:0} z:0xc00012c090} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:0} z:0xc00012c090} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] cap4:[] cap6:[] state:0xc000bf88b0 shieldsUp:false}"
2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine DERP map  derp_map="{\"Regions\":{\"1000\":{\"EmbeddedRelay\":false,\"RegionID\":1000,\"RegionCode\":\"coder_stun_1000\",\"RegionName\":\"Coder STUN 1000\",\"Nodes\":[{\"Name\":\"1000stun0\",\"RegionID\":1000,\"HostName\":\"stun.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1001\":{\"EmbeddedRelay\":false,\"RegionID\":1001,\"RegionCode\":\"coder_stun_1001\",\"RegionName\":\"Coder STUN 1001\",\"Nodes\":[{\"Name\":\"1001stun0\",\"RegionID\":1001,\"HostName\":\"stun1.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1002\":{\"EmbeddedRelay\":false,\"RegionID\":1002,\"RegionCode\":\"coder_stun_1002\",\"RegionName\":\"Coder STUN 1002\",\"Nodes\":[{\"Name\":\"1002stun0\",\"RegionID\":1002,\"HostName\":\"stun2.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1003\":{\"EmbeddedRelay\":false,\"RegionID\":1003,\"RegionCode\":\"coder_stun_1003\",\"RegionName\":\"Coder STUN 1003\",\"Nodes\":[{\"Name\":\"1003stun0\",\"RegionID\":1003,\"HostName\":\"stun3.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1004\":{\"EmbeddedRelay\":false,\"RegionID\":1004,\"RegionCode\":\"coder_stun_1004\",\"RegionName\":\"Coder STUN 1004\",\"Nodes\":[{\"Name\":\"1004stun0\",\"RegionID\":1004,\"HostName\":\"stun4.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"999\":{\"EmbeddedRelay\":true,\"RegionID\":999,\"RegionCode\":\"coder\",\"RegionName\":\"Coder Embedded Relay\",\"Nodes\":[{\"Name\":\"999b\",\"RegionID\":999,\"HostName\":\"https://coder.<domain.com>\",\"STUNPort\":-1,\"DERPPort\":443}]}}}"
2024-08-23 04:24:01.061 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-0
2024-08-23 04:24:01.061 [info]  provisionerd-coder-79c885844-z5p7q-0: successfully connected to coderd
2024-08-23 04:24:01.082 [info]  dbpurge: purged old database entries  duration=89.170548ms
2024-08-23 04:24:01.146 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-2
2024-08-23 04:24:01.146 [info]  provisionerd-coder-79c885844-z5p7q-2: successfully connected to coderd
2024-08-23 04:24:01.255 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-1
2024-08-23 04:24:01.255 [info]  provisionerd-coder-79c885844-z5p7q-1: successfully connected to coderd
2024-08-23 04:24:01.747 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
2024-08-23 04:24:02.052 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
2024-08-23 04:24:02.058 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
```

Looks like it eventually did come up.

We are installing coder on bare metal on-prem Kubernetes cluster.

Below is the yaml that I used:

```yaml
coder:
  env:
    - name: CODER_PG_CONNECTION_URL
      valueFrom:
        secretKeyRef:
          name: coder-db-url
          key: url

    - name: CODER_ACCESS_URL
      value: "https://coder.<domain.com>"

  service:
    type: NodePort
    sessionAffinity: None
    externalTrafficPolicy: Local
```

TL;DR -- 
1) `2024-08-23 04:24:01.082 [info]  dbpurge: purged old database entries  duration=89.170548ms` indicates that it took coder sometime to setup the initial database
2) But probe timeout of `1-sec` is probably too aggressive if you are just starting up coder.   Probes would likely always fail
3) On bare metal setups (w/ calico), looks like coder also experienced timeout connecting to Postgres the first time.  Upon restart it did connect successfully.

Unfortunately, I don't see a way to override probe timeout values via `values.yaml`.

**Comment by Lolelinus13** [comment]
> jFYI, I tried a few things and here are my observations:
> 
>     1. I disabled probes and started to see following logs
> 
> 
> ```
> Started HTTP listener at http://0.0.0.0:8080
> 
> View the Web UI: https://coder.<domain.com>
> 2024-08-23 04:16:21.718 [erro]  connect to postgres failed ...
>     error= migrate up:
>                github.com/coder/coder/v2/cli.ConnectToPostgres
>                    /home/runner/work/coder/coder/cli/server.go:2157
>              - up:
>                github.com/coder/coder/v2/coderd/database/migrations.UpWithFS
>                    /home/runner/work/coder/coder/coderd/database/migrations/migrate.go:82
>              - timeout: can't acquire database lock
> Encountered an error running "coder server", see "coder server --help" for more information
> error: connect to postgres: connect to postgres: migrate up: up: timeout: can't acquire database lock
> ```
> 
>     2. It crashed and restarted, then I saw these logs
> 
> 
> ```
> Started HTTP listener at http://0.0.0.0:8080
> 
> View the Web UI: https://coder.<domain.com>
> 2024-08-23 04:23:59.471 [info]  pubsub: pubsub dialing postgres  network=tcp  address=postgres-coder-0.postgres-coder.coder.svc.cluster.local:5432  timeout_ms=0
> 2024-08-23 04:23:59.502 [info]  pubsub: pubsub postgres TCP connection established  network=tcp  address=postgres-coder-0.postgres-coder.coder.svc.cluster.local:5432  timeout_ms=0  elapsed_ms=30
> 2024-08-23 04:23:59.514 [info]  pubsub: pubsub connected to postgres
> 2024-08-23 04:23:59.514 [info]  pubsub: pubsub has started
> 
> ==> Logs will stream in below (press ctrl+c to gracefully exit):
> 2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine network map  network_map="netmap: self: [6O1/Y] auth=machine-unknown u=? [fd7a:115c:a1e0:4312:8465:85bc:3ed5:4a10/128]\n"
> 2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine filter  filter="&{logf:0x136d7c0 local:0xc000b7ebb8 logIPs:0xc000b7ebd0 matches4:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:281470681743360} z:0xc00012c078} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:281470681743360} z:0xc00012c078} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] matches6:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:0} z:0xc00012c090} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:0} z:0xc00012c090} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] cap4:[] cap6:[] state:0xc000bf88b0 shieldsUp:false}"
> 2024-08-23 04:24:00.511 [info]  coderd.servertailnet: updating engine DERP map  derp_map="{\"Regions\":{\"1000\":{\"EmbeddedRelay\":false,\"RegionID\":1000,\"RegionCode\":\"coder_stun_1000\",\"RegionName\":\"Coder STUN 1000\",\"Nodes\":[{\"Name\":\"1000stun0\",\"RegionID\":1000,\"HostName\":\"stun.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1001\":{\"EmbeddedRelay\":false,\"RegionID\":1001,\"RegionCode\":\"coder_stun_1001\",\"RegionName\":\"Coder STUN 1001\",\"Nodes\":[{\"Name\":\"1001stun0\",\"RegionID\":1001,\"HostName\":\"stun1.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1002\":{\"EmbeddedRelay\":false,\"RegionID\":1002,\"RegionCode\":\"coder_stun_1002\",\"RegionName\":\"Coder STUN 1002\",\"Nodes\":[{\"Name\":\"1002stun0\",\"RegionID\":1002,\"HostName\":\"stun2.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1003\":{\"EmbeddedRelay\":false,\"RegionID\":1003,\"RegionCode\":\"coder_stun_1003\",\"RegionName\":\"Coder STUN 1003\",\"Nodes\":[{\"Name\":\"1003stun0\",\"RegionID\":1003,\"HostName\":\"stun3.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"1004\":{\"EmbeddedRelay\":false,\"RegionID\":1004,\"RegionCode\":\"coder_stun_1004\",\"RegionName\":\"Coder STUN 1004\",\"Nodes\":[{\"Name\":\"1004stun0\",\"RegionID\":1004,\"HostName\":\"stun4.l.google.com\",\"STUNPort\":19302,\"STUNOnly\":true}]},\"999\":{\"EmbeddedRelay\":true,\"RegionID\":999,\"RegionCode\":\"coder\",\"RegionName\":\"Coder Embedded Relay\",\"Nodes\":[{\"Name\":\"999b\",\"RegionID\":999,\"HostName\":\"https://coder.<domain.com>\",\"STUNPort\":-1,\"DERPPort\":443}]}}}"
> 2024-08-23 04:24:01.061 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-0
> 2024-08-23 04:24:01.061 [info]  provisionerd-coder-79c885844-z5p7q-0: successfully connected to coderd
> 2024-08-23 04:24:01.082 [info]  dbpurge: purged old database entries  duration=89.170548ms
> 2024-08-23 04:24:01.146 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-2
> 2024-08-23 04:24:01.146 [info]  provisionerd-coder-79c885844-z5p7q-2: successfully connected to coderd
> 2024-08-23 04:24:01.255 [info]  coderd: starting in-memory provisioner daemon  name=coder-79c885844-z5p7q-1
> 2024-08-23 04:24:01.255 [info]  provisionerd-coder-79c885844-z5p7q-1: successfully connected to coderd
> 2024-08-23 04:24:01.747 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
> 2024-08-23 04:24:02.052 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
> 2024-08-23 04:24:02.058 [info]  terraform: detected terraform version  installed_version=1.8.4  min_version=1.1.0  max_version=1.8.9
> ```
> 
> Looks like it eventually did come up.
> 
> We are installing coder on bare metal on-prem Kubernetes cluster.
> 
> Below is the yaml that I used:
> 
> ```yaml
> coder:
>   env:
>     - name: CODER_PG_CONNECTION_URL
>       valueFrom:
>         secretKeyRef:
>           name: coder-db-url
>           key: url
> 
>     - name: CODER_ACCESS_URL
>       value: "https://coder.<domain.com>"
> 
>   service:
>     type: NodePort
>     sessionAffinity: None
>     externalTrafficPolicy: Local
> ```
> 
> TL;DR --
> 
>     1. `2024-08-23 04:24:01.082 [info]  dbpurge: purged old database entries  duration=89.170548ms` indicates that it took coder sometime to setup the initial database
> 
>     2. But probe timeout of `1-sec` is probably too aggressive if you are just starting up coder.   Probes would likely always fail
> 
>     3. On bare metal setups (w/ calico), looks like coder also experienced timeout connecting to Postgres the first time.  Upon restart it did connect successfully.
> 
> 
> Unfortunately, I don't see a way to override probe timeout values via `values.yaml`.
> 
> PS: Finally, I re-enabled probes and now everything is normal once DB setup/migration has finished

did you downloaded the chart and disabled the probes and then installed it or is their a possibility in the values to disable them?

**Comment by Cjkjvfnby** [comment]
Total timing for probes are shorter than in-service timeout for connecting to the DB.  

Coder pod is restarted before it prints error to the logs. 

To get my log messages I need to edit the probes manually. 

```
kubectl edit deployment coder -n coder
```

So I finally can see:
```
[erro]  connect to postgres failed ...
    error= unable to connect after 1 tries; last error:
               github.com/coder/coder/v2/cli.ConnectToPostgres
                   /home/runner/work/coder/coder/cli/server.go:2181
             - dial TCP <....>:5432: connect: connection timed out
```

It would be nice if timeout for PG will be aligned with the probes.
